include/RAJA/pattern/kernel/Lambda.hpp

```diff
@@ -79,6 +91,89 @@ struct StatementExecutor<statement::Lambda<LoopIndex>> {
   }
 };
 
+template<camp::idx_t Pos, typename... Args>
+struct inspector
+{};
+
+template<camp::idx_t Pos, camp::idx_t id, typename... Tail>
+struct inspector<Pos, RAJA::statement::Param<id>, Tail...>
+{
+  template<typename T, typename Data>
+  static auto make_args(T tuple, Data &&data) ->
+  decltype(inspector<Pos-1, Tail...>::make_args(camp::tuple_cat_pair(tuple, camp::make_tuple(camp::get<id>(data.param_tuple))), data))
+  {
+    auto newTuple = camp::tuple_cat_pair(tuple, camp::make_tuple(camp::get<id>(data.param_tuple)));
+    return inspector<Pos-1, Tail...>::make_args(newTuple, data);
+  }
+};
+
+template<camp::idx_t Pos, camp::idx_t id, typename...Tail>
+struct inspector<Pos, RAJA::statement::Seg<id>, Tail...>
+{
+  template<typename T, typename Data>
+  static auto make_args(T tuple, Data &&data) ->
+    decltype(inspector<Pos-1, Tail...>::make_args
+             (camp::tuple_cat_pair(tuple,camp::make_tuple(camp::get<id>(data.segment_tuple).begin()[camp::get<id>(data.offset_tuple)])), data))
+ {
+    auto newTuple = camp::tuple_cat_pair
+      (tuple,camp::make_tuple(camp::get<id>(data.segment_tuple).begin()[camp::get<id>(data.offset_tuple)]));
+
+    return inspector<Pos-1, Tail...>::make_args(newTuple, data);
+  }
+};
+
+template<camp::idx_t id>
+struct inspector<1, RAJA::statement::Param<id>>
+{
+  template<typename T, typename Data>
+  static auto make_args(T tuple, Data &&data) ->
+    decltype(camp::tuple_cat_pair(tuple, camp::make_tuple(camp::get<id>(data.param_tuple))))
+  {
+    return camp::tuple_cat_pair(tuple, camp::make_tuple(camp::get<id>(data.param_tuple)));
+  }
+};
+
+template<camp::idx_t id>
+struct inspector<1, RAJA::statement::Seg<id>>
+{
+  template<typename T, typename Data>
+    static auto make_args(T tuple, Data &&data) ->
+    decltype(camp::tuple_cat_pair(tuple,camp::make_tuple(camp::get<id>(data.segment_tuple).begin()[camp::get<id>(data.offset_tuple)])))
+  {
+    return camp::tuple_cat_pair
+      (tuple,camp::make_tuple(camp::get<id>(data.segment_tuple).begin()[camp::get<id>(data.offset_tuple)]));
+  }
+};
+
+//Iteration 1
+template <camp::idx_t LoopIndex,typename... Args>
+struct StatementExecutor<statement::Lambda<LoopIndex, Args...>> {
+
+  template <typename Data>
+  static RAJA_INLINE void exec(Data &&data)
+  {
+    const int size = sizeof...(Args); //Total number of arguments
+    auto myTuple = inspector<size, Args...>::make_args(camp::make_tuple(),data);
```

Comment:
In a nutshell, I'm doing a gather of the desired arguments and their specified order, storing them in a tuple then feeding them into the lambda. I'm wondering if there is a cleaner way to do this...

Reply:
it might be cleaner to do something that looked like:

auto myTuple = camp::make_tuple( inspector<Args>(data)::extract_arg() ... );

Where you use the helper class (inspector) to extract each argument as a single value... this eliminates the need for peeling off each argument and using tuple_cat_pair.
It would also make the implementation a lot simpler.

Reply:
sorry, "data" above should be the argument to extract_arg, like this:

`auto myTuple = camp::make_tuple( inspector<Args>::extract_arg(data) ... );`

Reply:
Sweet! It works. I'll do some clean up on this branch next.